### PR TITLE
Feat/expires

### PR DESCRIPTION
**Pull request by rudyardrichter**
### New Features

- Allow specifying an `expires_at` field to expire a user's access via policies or group membership.

### CODE REVIEW

File: arborist/server.go [hunk 1]
```diff
@@ -972,7 +974,22 @@ func (server *Server) handleUserGrantPolicy(w http.ResponseWriter, r *http.Reque
 		_ = response.write(w, r)
 		return
 	}
-	errResponse := grantUserPolicy(server.db, username, requestPolicy.PolicyName)
+	var expiresAt *time.Time
+	if requestPolicy.ExpiresAt != "" {
+		exp, err := time.Parse(time.RFC3339, requestPolicy.ExpiresAt)
+		if err != nil {
```
Comment (Avantol13) [inline arborist/server.go]: is this double parse necessary? Can you just attempt the parse to `time.RFC3339Nano`?

File: arborist/server.go [hunk 2]
```diff
@@ -1274,7 +1292,18 @@ func (server *Server) handleGroupAddUser(w http.ResponseWriter, r *http.Request,
 		_ = response.write(w, r)
 		return
 	}
-	errResponse := addUserToGroup(server.db, requestUser.Username, groupName)
+	var expiresAt time.Time
+	if requestUser.ExpiresAt != "" {
+		expiresAt, err = time.Parse(time.RFC3339, requestUser.ExpiresAt)
```
Comment (Avantol13) [inline arborist/server.go]: why no attempted parse to `Nano` here?

File: docs/openapi.yaml [hunk 3]
```diff
@@ -964,6 +964,33 @@ paths:
...
           description: group not found
+  /group/{groupName}/user:
+    parameters:
+      - in: path
+        name: groupName
+        required: true
+        schema:
+          type: string
+        description: the groupName for a group registered in arborist
+    post:
```
[the remark this answers is not in the document]
Comment (rudyardrichter) [reply docs/openapi.yaml]: I'm a confused what you mean, I think this is the desired behavior right? `paramaters` here means path parameters not JSON
image

File: arborist/server.go (hunk 1 above)
Comment (rudyardrichter) [reply arborist/server.go]: honestly nano should not be necessary, I'll take that out

File: arborist/server.go (hunk 2 above)
Comment (rudyardrichter) [reply arborist/server.go]: see above, going to skip nano

File: docs/openapi.yaml [hunk 4]
```diff
@@ -964,6 +964,33 @@ paths:
           description: successfully revoked
         404:
           description: group not found
+  /group/{groupName}/user:
```
[the remark this answers is not in the document]
Comment (rudyardrichter) [reply docs/openapi.yaml]: will update!

[the remark this answers is not in the document]
Comment (rudyardrichter) [reply docs/openapi.yaml]: updated!

File: docs/openapi.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (Avantol13) [reply docs/openapi.yaml]: the paramaters are not only path, they're JSON body too

[the remark this answers is not in the document]
Comment (Avantol13) [reply docs/openapi.yaml]: :man_facepalming: yup, missed the request body section somehow

File: arborist/user.go [hunk 5]
```diff
@@ -200,18 +227,18 @@ func revokeUserPolicyAll(db *sqlx.DB, username string) *ErrorResponse {
...
 	}
 	if username == "" {
 		return newErrorResponse("missing `username` argument", 400, nil)
 	}
 	stmt := `
-		INSERT INTO usr_grp(usr_id, grp_id)
-		VALUES ((SELECT id FROM usr WHERE name = $1), (SELECT id FROM grp WHERE name = $2))
+		INSERT INTO usr_grp(usr_id, grp_id, expires_at)
+		VALUES ((SELECT id FROM usr WHERE name = $1), (SELECT id FROM grp WHERE name = $2), $3)
 	`
```
Comment (Avantol13) [inline arborist/user.go]: adding `ON CONFLICT (usr_id, grp_id) DO UPDATE SET expires_at = EXCLUDED.expires_at` should be safe here?

**Pull request merged**